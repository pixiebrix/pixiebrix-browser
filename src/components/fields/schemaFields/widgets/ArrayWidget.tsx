/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { SchemaFieldProps } from "@/components/fields/schemaFields/propTypes";
import { FieldArray, useField } from "formik";
import { fieldLabel } from "@/components/fields/fieldUtils";
import { Button } from "react-bootstrap";
import React from "react";
import { Schema } from "@/core";
import {
  booleanPredicate,
  findOneOf,
  textPredicate,
} from "@/components/fields/schemaFields/schemaUtils";
import { UnknownObject } from "@/types";
import { defaultBlockConfig } from "@/blocks/util";
import SchemaField from "@/components/fields/schemaFields/SchemaField";
import { joinName } from "@/utils";

// Empty value for text fields for the Formik state
const EMPTY_TEXT_VALUE = "";

function getDefaultArrayItem(schema: Schema): unknown {
  // TODO: handle enum and const

  if (schema.default) {
    return schema.default;
  }

  if (textPredicate(schema)) {
    return EMPTY_TEXT_VALUE;
  }

  if (schema.type === "object") {
    return defaultBlockConfig(schema);
  }

  if (findOneOf(schema, booleanPredicate)) {
    return false;
  }

  if (findOneOf(schema, textPredicate)) {
    return EMPTY_TEXT_VALUE;
  }

  return null;
}

const ArrayWidget: React.FC<SchemaFieldProps<unknown>> = ({
  schema,
  label,
  name,
}) => {
  const [field] = useField<UnknownObject[]>(name);

  if (Array.isArray(schema.items)) {
    throw new TypeError("Support for arrays of mixed types is not implemented");
  }

  if (typeof schema.items === "boolean") {
    throw new TypeError("Schema required for items");
  }

  const schemaItems = schema.items as Schema;

  return (
    <FieldArray name={name}>
      {({ remove, push }) => (
        <>
          <ul className="list-group">
            {(field.value ?? []).map((item: unknown, index: number) => (
              <li className="list-group-item" key={index}>
                <SchemaField
                  key={index}
                  name={joinName(name, String(index))}
                  schema={schemaItems}
                  label={
                    <span>
                      {label ?? fieldLabel(name)} #{index + 1}
                    </span>
                  }
                />
                <Button
                  variant="danger"
                  size="sm"
                  onClick={() => {
                    remove(index);
                  }}
                >
                  Remove Item
                </Button>
              </li>
            ))}
          </ul>
          <Button
            onClick={() => {
              push(getDefaultArrayItem(schemaItems));
            }}
          >
            Add Item
          </Button>
        </>
      )}
    </FieldArray>
  );
};

export default ArrayWidget;
