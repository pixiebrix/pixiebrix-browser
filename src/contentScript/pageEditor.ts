/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { serializeError } from "serialize-error";
import {
  blockReducer,
  type IntermediateState,
  type ReduceOptions,
} from "@/runtime/reducePipeline";
import { type BlockConfig } from "@/blocks/types";
import { cloneDeep } from "lodash";
import ConsoleLogger from "@/utils/ConsoleLogger";
import { type SerializableResponse } from "@/pageScript/messenger/pigeon";
import apiVersionOptions from "@/runtime/apiVersionOptions";
import { $safeFind } from "@/helpers";
import { clearDynamicElements } from "@/contentScript/pageEditor/dynamic";
import { reactivateTab } from "./lifecycle";
import {
  BusinessError,
  CancelError,
  NoRendererError,
} from "@/errors/businessErrors";
import { uuidv4 } from "@/types/helpers";
import { type PanelPayload } from "@/sidebar/types";
import { HeadlessModeError } from "@/blocks/errors";
import { showTemporarySidebarPanel } from "@/contentScript/sidebarController";
import { stopInspectingNativeHandler } from "./pageEditor/elementPicker";
import { showModal } from "@/blocks/transformers/ephemeralForm/modalUtils";
import { createFrameSource } from "@/blocks/transformers/temporaryInfo/DisplayTemporaryInfo";
import { waitForTemporaryPanel } from "@/blocks/transformers/temporaryInfo/temporaryPanelProtocol";
import { type ApiVersion, type BlockArgsContext } from "@/types/runtimeTypes";
import { type UUID } from "@/types/stringTypes";
import { type RegistryId } from "@/types/registryTypes";

export type RunBlockArgs = {
  /**
   * The runtime API version to use
   */
  apiVersion: ApiVersion;
  /**
   * The IBlock configuration.
   */
  blockConfig: BlockConfig;
  /**
   * Context to render the BlockArg, should include @input, @options, and service context
   * @see makeServiceContext
   */
  context: BlockArgsContext;
  /**
   * Root jQuery selector to determine the root if the rootMode is "inherit".
   * @see BlockConfig.rootMode
   */
  rootSelector: string | undefined;
};

/**
 * Run a single block (e.g., for generating output previews)
 * @see BlockPreview
 */
export async function runBlock({
  blockConfig,
  context,
  apiVersion,
  rootSelector,
}: RunBlockArgs): Promise<unknown> {
  const versionOptions = apiVersionOptions(apiVersion);

  if (!versionOptions.explicitDataFlow) {
    throw new BusinessError(
      "Preview only supported for mods using runtime v2 or later"
    );
  }

  const state: IntermediateState = {
    context,
    // Can pick any index. It's only used for adding context to log messages, and we're disabling value logging
    // below with `logValues: false`
    index: 0,
    // Force isLastBlock so blockReducer does not complain about the outputKey being forced to undefined
    isLastBlock: true,
    // `root` is over-ridden below if rootSelector is provided
    root: document,
    // We're forcing apiVersion: 2 or higher above values must come from the context
    previousOutput: {},
  };

  if (rootSelector) {
    // Handle non-document contexts. If the selector is unique, this gives the root that would be available at runtime.
    // Differences in behavior:
    // - For triggers, the PixieBrix looks for the closest ancestor to the DOM event target matching the selector
    //   See TriggerExtensionPoint.eventHandler for reference
    // - For multi-menus (not currently available in the Page Editor), the below logic returns an arbitrary menu
    const rootElement = $safeFind(rootSelector);
    if (rootElement.length > 0) {
      state.root = rootElement.get(0);
    }
  }

  const options: ReduceOptions = {
    ...versionOptions,
    branches: [],
    headless: true,
    logValues: false,
    logger: new ConsoleLogger(),
    // Excluding runId will prevent the run from being stored in traces
    runId: null,
    extensionId: null,
  };

  // Exclude the outputKey so that `output` is the output of the brick. Alternatively we could have taken then
  // value from the context[outputKey] from the return value of blockReducer
  const { output } = await blockReducer(
    { ...blockConfig, outputKey: undefined },
    state,
    options
  );

  return cloneDeep(output) as SerializableResponse;
}

type Location = "modal" | "panel";

/**
 * Run a single renderer (e.g. - for running a block preview)
 *
 * Renderers need to be run with try-catch, catch the HeadlessModeError, and
 * use that to send the panel payload to the sidebar (or other target)
 * @see SidebarExtensionPoint
 *  starting on line 184, the call to reduceExtensionPipeline(),
 *  wrapped in a try-catch
 * @see executeBlockWithValidatedProps
 *  starting on line 323, the runRendererPipeline() function
 *
 * Note: Currently only implemented for the temporary sidebar panels
 * @see useDocumentPreviewRunBlock
 */
export async function runRendererBlock({
  extensionId,
  blueprintId,
  runId,
  title,
  args,
  location,
}: {
  extensionId: UUID;
  blueprintId: RegistryId | null;
  runId: UUID;
  title: string;
  args: RunBlockArgs;
  location: Location;
}): Promise<void> {
  const nonce = uuidv4();

  let payload: PanelPayload;
  try {
    await runBlock(args);
    // We're expecting a HeadlessModeError (or other error) to be thrown in the line above
    // noinspection ExceptionCaughtLocallyJS
    throw new NoRendererError();
  } catch (error) {
    if (error instanceof HeadlessModeError) {
      payload = {
        key: nonce,
        blockId: error.blockId,
        args: error.args,
        ctxt: error.ctxt,
        extensionId,
        runId,
      };
    } else {
      payload = {
        key: nonce,
        error: serializeError(error),
        extensionId,
        runId,
      };
    }

    if (location === "panel") {
      showTemporarySidebarPanel({
        // Pass extension id so previous run is cancelled
        extensionId,
        blueprintId,
        nonce,
        heading: title,
        payload,
      });
    } else if (location === "modal") {
      const controller = new AbortController();
      const url = await createFrameSource(nonce, "modal");

      showModal({ url, controller });

      try {
        await waitForTemporaryPanel(nonce, {
          extensionId,
          blueprintId,
          nonce,
          heading: title,
          payload,
        });
      } catch (error) {
        // Match behavior of Display Temporary Info
        if (error instanceof CancelError) {
          // NOP
        } else {
          throw error;
        }
      } finally {
        controller.abort();
      }
    } else {
      // eslint-disable-next-line @typescript-eslint/restrict-template-expressions -- dynamic check for never
      throw new Error(`Support for previewing in ${location} not implemented`);
    }
  }
}

export async function resetTab(): Promise<void> {
  stopInspectingNativeHandler();
  await clearDynamicElements({});
  await reactivateTab();
}
