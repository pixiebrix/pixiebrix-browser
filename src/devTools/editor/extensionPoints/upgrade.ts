/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { BlockConfig, BlockPipeline } from "@/blocks/types";
import blockRegistry from "@/blocks/registry";
import { inputProperties } from "@/helpers";
import { UnknownObject } from "@/types";
import { Expression, SchemaDefinition, TemplateEngine } from "@/core";
import { cloneDeep } from "lodash";

export function stringToExpression(
  value: unknown,
  templateEngine: TemplateEngine
): Expression {
  if (typeof value !== "string") {
    throw new TypeError("Expected string for value");
  }

  const type = /^@\S+$/g.test(value) ? "var" : templateEngine;
  return {
    __type__: type,
    __value__: value,
  };
}

async function upgradeBlock(blockConfig: BlockConfig): Promise<void> {
  const { inputSchema } = await blockRegistry.lookup(blockConfig.id);
  const inputProps = inputProperties(inputSchema);

  // `v1`/`v2` use mustache by default
  const templateEngine = blockConfig.templateEngine ?? "mustache";

  await Promise.all(
    Object.keys(blockConfig.config).map(async (name) => {
      await upgradeValue(
        blockConfig.config,
        name,
        // eslint-disable-next-line security/detect-object-injection -- name comes from config keys
        inputProps[name],
        templateEngine
      );
    })
  );

  if (blockConfig.if) {
    blockConfig.if = stringToExpression(blockConfig.if, templateEngine);
  }

  // Top-level `templateEngine` not supported in v3
  delete blockConfig.templateEngine;
}

async function upgradeValue(
  config: UnknownObject,
  fieldName: string,
  fieldSchema: SchemaDefinition,
  templateEngine: TemplateEngine
): Promise<void> {
  if (typeof fieldSchema === "boolean") {
    return;
  }

  // eslint-disable-next-line security/detect-object-injection -- caller iterates over keys
  const value = config[fieldName];
  if (typeof value === "object" && fieldSchema) {
    if (
      Array.isArray(value) &&
      fieldSchema.type === "array" &&
      typeof fieldSchema.items !== "boolean"
    ) {
      const items: SchemaDefinition[] = [];
      const { additionalItems } = fieldSchema;
      const additionalItemsByType: Record<string, SchemaDefinition> = {};

      if (Array.isArray(fieldSchema.items)) {
        items.push(...fieldSchema.items);
      } else if (fieldSchema.items) {
        items.push(fieldSchema.items);
        if (typeof fieldSchema.items.type === "string") {
          additionalItemsByType[fieldSchema.items.type] = fieldSchema.items;
        }
      }

      if (typeof additionalItems !== "boolean" && additionalItems) {
        if (typeof additionalItems.type === "string") {
          additionalItemsByType[additionalItems.type] = additionalItems;
        }

        if (additionalItems.oneOf) {
          for (const prop of additionalItems.oneOf) {
            if (typeof prop !== "boolean" && typeof prop.type === "string") {
              additionalItemsByType[prop.type] = prop;
            }
          }
        }
      }

      for (const [index, element] of value.entries()) {
        if (index >= items.length) {
          // eslint-disable-next-line security/detect-object-injection
          items[index] = additionalItemsByType[typeof element];
        }
      }

      await Promise.all(
        value.map(async (element, index) => {
          await upgradeValue(
            (value as unknown) as UnknownObject,
            index.toString(),
            // eslint-disable-next-line security/detect-object-injection
            items[index],
            templateEngine
          );
        })
      );
    } else if (fieldSchema.type === "object") {
      const properties = fieldSchema.properties ?? {};

      if (
        typeof fieldSchema.additionalProperties !== "boolean" &&
        fieldSchema.additionalProperties
      ) {
        const additionalPropsByType: Record<string, SchemaDefinition> = {};
        if (typeof fieldSchema.additionalProperties.type === "string") {
          additionalPropsByType[fieldSchema.additionalProperties.type] =
            fieldSchema.additionalProperties;
        }

        if (fieldSchema.additionalProperties.oneOf) {
          for (const prop of fieldSchema.additionalProperties.oneOf) {
            if (typeof prop !== "boolean" && typeof prop.type === "string") {
              additionalPropsByType[prop.type] = prop;
            }
          }
        }

        for (const name of Object.keys(value).filter(
          (key) => !(key in properties)
        )) {
          // eslint-disable-next-line security/detect-object-injection
          const subValue = (value as UnknownObject)[name];
          // eslint-disable-next-line security/detect-object-injection
          properties[name] = additionalPropsByType[typeof subValue];
        }
      }

      await Promise.all(
        Object.keys(value).map(async (name) => {
          await upgradeValue(
            value as UnknownObject,
            name,
            // eslint-disable-next-line security/detect-object-injection
            properties[name],
            templateEngine
          );
        })
      );
    }
  } else if (
    fieldSchema?.type === "string" &&
    fieldSchema?.format === "selector"
  ) {
    // NOP: the Page Editor doesn't support templated selectors
  } else if (typeof value === "string") {
    // eslint-disable-next-line security/detect-object-injection -- caller iterates over keys
    config[fieldName] = stringToExpression(value, templateEngine);
  }
}

/**
 * Attempt to upgrade the blocks in a pipeline from api v2 to v3
 */
export async function upgradePipelineToV3(
  blockPipeline: BlockPipeline
): Promise<BlockPipeline> {
  const cloned = cloneDeep(blockPipeline);
  await Promise.all(
    cloned.map(async (block) => {
      await upgradeBlock(block);
    })
  );
  return cloned;
}
