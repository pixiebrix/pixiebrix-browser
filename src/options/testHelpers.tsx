/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { configureStore } from "@reduxjs/toolkit";
import { authSlice } from "@/auth/authSlice";
import extensionsSlice from "@/store/extensionsSlice";
import settingsSlice from "@/store/settingsSlice";
import { blueprintModalsSlice } from "@/options/pages/blueprints/modals/blueprintModalsSlice";
import { createRenderWithWrappers } from "@/testUtils/testHelpers";
import blueprintsSlice from "@/options/pages/blueprints/blueprintsSlice";
import { recipesSlice } from "@/recipes/recipesSlice";

const renderWithWrappers = createRenderWithWrappers(() =>
  configureStore({
    reducer: {
      auth: authSlice.reducer,
      settings: settingsSlice.reducer,
      options: extensionsSlice.reducer,
      blueprintModals: blueprintModalsSlice.reducer,
      blueprints: blueprintsSlice.reducer,
      recipes: recipesSlice.reducer,
    },
  })
);

// eslint-disable-next-line import/export -- re-export RTL
export * from "@testing-library/react";
// eslint-disable-next-line import/export -- override render
export { renderWithWrappers as render };
