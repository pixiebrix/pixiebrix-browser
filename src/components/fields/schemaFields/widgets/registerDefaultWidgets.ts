/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import UrlMatchPatternWidget from "./UrlMatchPatternWidget";
import UrlPatternWidget from "@/pageEditor/components/UrlPatternWidget";
import ArrayWidget from "./ArrayWidget";
import BooleanWidget from "./BooleanWidget";
import IntegerWidget from "./IntegerWidget";
import NumberWidget from "./NumberWidget";
import ObjectWidget from "./ObjectWidget";
import OmitFieldWidget from "./OmitFieldWidget";
import SchemaSelectWidget from "./SchemaSelectWidget";
import TemplateToggleWidget from "./TemplateToggleWidget";
import TextWidget from "./TextWidget";
import UnsupportedWidget from "./UnsupportedWidget";
import widgetsRegistry from "./widgetsRegistry";
import WorkshopMessageWidget from "./WorkshopMessageWidget";
import SelectorMatchWidget from "@/pageEditor/components/SelectorMatchWidget";
import CssClassWidget from "./CssClassWidget";
import DatabaseWidget from "./DatabaseWidget";
import PasswordWidget from "./PasswordWidget";
import IconWidget from "@/components/fields/IconWidget";

const defaultWidgets = {
  ArrayWidget,
  BooleanWidget,
  IntegerWidget,
  NumberWidget,
  ObjectWidget,
  OmitFieldWidget,
  SchemaSelectWidget,
  TemplateToggleWidget,
  TextWidget,
  UnsupportedWidget,
  UrlMatchPatternWidget,
  UrlPatternWidget,
  SelectorMatchWidget,
  CssClassWidget,
  WorkshopMessageWidget,
  DatabaseWidget,
  IconWidget,
  PasswordWidget,
} as const;

function registerDefaultWidgets() {
  Object.assign(widgetsRegistry, defaultWidgets);
  for (const [name, widget] of Object.entries(defaultWidgets)) {
    if (!widget) {
      throw new Error(
        `Error registering default widgets. ${name} is undefined. Is there a circular dependency?`
      );
    }
  }
}

export default registerDefaultWidgets;
