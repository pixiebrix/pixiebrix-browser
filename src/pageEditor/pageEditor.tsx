/*
 * Copyright (C) 2024 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import "@/vendors/bootstrapWithoutRem.css";
import "@/utils/global.scss";
import "@/utils/layout.scss";

import "@/extensionContext";
import "@/development/darkMode.js";

import { initMessengerLogging } from "@/development/messengerLogging";

import ReactDOM from "react-dom";
import React from "react";
import Panel from "@/pageEditor/Panel";
import { watchNavigation } from "@/pageEditor/protocol";
import { initToaster } from "@/utils/notify";
import { initRuntimeLogging } from "@/development/runtimeLogging";
import { initPerformanceMonitoring } from "@/telemetry/performance";
import { setPlatform } from "@/platform/platformContext";
import extensionPagePlatform from "@/extensionPages/extensionPagePlatform";

async function init() {
  setPlatform(extensionPagePlatform);
  void initMessengerLogging();
  void initRuntimeLogging();
  try {
    await initPerformanceMonitoring();
  } catch (error) {
    console.error("Failed to initialize performance monitoring", error);
  }

  watchNavigation();
  initToaster();

  ReactDOM.render(<Panel />, document.querySelector("#container"));
}

void init();
