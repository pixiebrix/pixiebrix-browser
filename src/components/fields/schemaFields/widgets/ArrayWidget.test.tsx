/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import registerDefaultWidgets from "@/components/fields/schemaFields/widgets/registerDefaultWidgets";
import { type Schema } from "@/core";
import ArrayWidget from "@/components/fields/schemaFields/widgets/ArrayWidget";
import userEvent from "@testing-library/user-event";
import { stringToExpression } from "@/pageEditor/extensionPoints/upgrade";
import { render, screen } from "@/pageEditor/testHelpers";

const fieldName = "testField";
const fieldDescription = "this is a test field description";

describe("ArrayWidget", () => {
  beforeAll(() => {
    registerDefaultWidgets();
  });

  test("renders empty widget", () => {
    const schema: Schema = {
      type: "array",
      additionalItems: true,
    };
    expect(
      render(
        <ArrayWidget
          name={fieldName}
          schema={schema}
          isRequired
          description={fieldDescription}
          defaultType="array"
        />,
        {
          initialValues: {
            [fieldName]: [],
          },
        }
      ).asFragment()
    ).toMatchSnapshot();
  });

  test("renders widget with items", () => {
    const schema: Schema = {
      type: "array",
      additionalItems: true,
      items: {
        type: "string",
      },
    };
    expect(
      render(
        <ArrayWidget
          name={fieldName}
          schema={schema}
          isRequired
          description={fieldDescription}
          defaultType="array"
        />,
        {
          initialValues: {
            [fieldName]: [
              "abc",
              "def",
              "this is a really long string to see what happens when the array value is long",
            ],
          },
        }
      ).asFragment()
    ).toMatchSnapshot();
  });

  test("add item", async () => {
    const schema: Schema = {
      type: "array",
      additionalItems: true,
    };
    const { getFormState } = render(
      <ArrayWidget
        name={fieldName}
        schema={schema}
        isRequired
        description={fieldDescription}
        defaultType="array"
      />,
      {
        initialValues: {
          [fieldName]: [],
        },
      }
    );

    // Add an item
    await userEvent.click(screen.getByText(/add item/i));

    const itemInput = screen.getByRole("textbox");

    // Change the item value
    await userEvent.type(itemInput, "myValue");

    const formState = await getFormState();

    expect(formState).toStrictEqual({
      [fieldName]: [stringToExpression("myValue", "nunjucks")],
    });
  });

  test("remove item", async () => {
    const schema: Schema = {
      type: "array",
      additionalItems: true,
      items: {
        type: "string",
      },
    };
    const { getFormState } = render(
      <ArrayWidget
        name={fieldName}
        schema={schema}
        isRequired
        description={fieldDescription}
        defaultType="array"
      />,
      {
        initialValues: {
          [fieldName]: ["abc", "def"],
        },
      }
    );

    // Open the field type toggle
    await userEvent.click(
      screen.getByTestId(`toggle-${fieldName}.1`).querySelector("button")
    );

    // Select "Remove" since we're in an array
    await userEvent.click(screen.getByText("Remove"));

    const formState = await getFormState();

    // Expect excluded property to be removed from the state
    expect(formState).toStrictEqual({
      [fieldName]: ["abc"],
    });
  });
});
