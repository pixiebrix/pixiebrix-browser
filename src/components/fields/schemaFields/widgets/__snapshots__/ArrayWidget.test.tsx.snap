// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArrayWidget renders empty widget 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <ul
      class="list-group mb-2"
    />
    <button
      class="btn btn-primary"
      type="button"
    >
      Add Item
    </button>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`ArrayWidget renders widget with items 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <ul
      class="list-group mb-2"
    >
      <li
        class="list-group-item py-1"
      >
        <div
          class="formGroup mb-0 form-group row"
        >
          <div
            class="col-xl-12 col-lg-12"
          >
            <div
              class="root"
            >
              <div
                class="field"
              >
                <textarea
                  class="form-control"
                  id="testField.0"
                  name="testField.0"
                  rows="1"
                >
                  abc
                </textarea>
              </div>
              <div
                class="dropdown dropdown"
                data-test-selected="Text"
                data-testid="toggle-testField.0"
              >
                <button
                  aria-expanded="false"
                  aria-haspopup="true"
                  class="dropdown-toggle btn btn-link"
                  type="button"
                >
                  <span
                    class="symbol"
                  >
                    <svg
                      class="root"
                    />
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </li>
      <li
        class="list-group-item py-1"
      >
        <div
          class="formGroup mb-0 form-group row"
        >
          <div
            class="col-xl-12 col-lg-12"
          >
            <div
              class="root"
            >
              <div
                class="field"
              >
                <textarea
                  class="form-control"
                  id="testField.1"
                  name="testField.1"
                  rows="1"
                >
                  def
                </textarea>
              </div>
              <div
                class="dropdown dropdown"
                data-test-selected="Text"
                data-testid="toggle-testField.1"
              >
                <button
                  aria-expanded="false"
                  aria-haspopup="true"
                  class="dropdown-toggle btn btn-link"
                  type="button"
                >
                  <span
                    class="symbol"
                  >
                    <svg
                      class="root"
                    />
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </li>
      <li
        class="list-group-item py-1"
      >
        <div
          class="formGroup mb-0 form-group row"
        >
          <div
            class="col-xl-12 col-lg-12"
          >
            <div
              class="root"
            >
              <div
                class="field"
              >
                <textarea
                  class="form-control"
                  id="testField.2"
                  name="testField.2"
                  rows="1"
                >
                  this is a really long string to see what happens when the array value is long
                </textarea>
              </div>
              <div
                class="dropdown dropdown"
                data-test-selected="Text"
                data-testid="toggle-testField.2"
              >
                <button
                  aria-expanded="false"
                  aria-haspopup="true"
                  class="dropdown-toggle btn btn-link"
                  type="button"
                >
                  <span
                    class="symbol"
                  >
                    <svg
                      class="root"
                    />
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </li>
    </ul>
    <button
      class="btn btn-primary"
      type="button"
    >
      Add Item
    </button>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
